### PR TITLE
Unify linker argument order across platforms

### PR DESCRIPTION
**Pull request by keith**
Previously extra linker arguments had different behavior on darwin vs
other unix platforms. On darwin the arguments passed with -Xlinker would
be passed to the linker before the default arguments, where as with the
default unix toolchain they would be passed afterwards.

There isn't really a great option for which order these should be in.
If you want to have a custom rpath that takes precedence over the
default rpaths, you want them to be passed before, but if you want to
negate a default argument you want them to come after.

This change unifies the behavior so at least you always get the same
behavior across platforms.

Resolves [SR-6508](https://bugs.swift.org/browse/SR-6508).

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
@jrose-apple can you take a look? or suggest who would be the right person to review this change?

**Comment by jrose-apple** [comment]
I'm the right person. All -X options are supposed to go after regular options, not before. Maybe we'll come up with an "-X<foo>-before" variant later, but Darwin is the designed behavior.

**Comment by keith** [comment]
> All -X options are supposed to go after regular options, not before.
>  but Darwin is the designed behavior.

On Darwin they appear to come before now, are you saying I should move those instead?

**Comment by jrose-apple** [comment]
Aargh, I can't read. On both toolchains they are in the middle but Darwin is close to the start and Unix is close to the end. They should be last-but-just-before-the-`-o`.

(All the other -X options probably need to be checked too, and then get a big comment making sure people don't move them.)

**Comment by keith** [comment]
Sounds good I'll move these down, add the comment, fix the tests, and look for other cases for a separate change

**Comment by keith** [comment]
I've moved these down for both the unix and darwin toolchains

**Comment by keith** [comment]
I don't see any other `-X` options in these files (assuming they follow the same naming conventions), were there specific ones you were thinking about? Or did you mean ones somewhere else in the codebase?

**Review by jrose-apple — COMMENTED** [review]
Sorry, I realized there's one more problem with the existing code that ought to be fixed while you're here!

File: lib/Driver/UnixToolChains.cpp [hunk 1]
```diff
@@ -325,6 +322,10 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
     Arguments.push_back("-v");
   }
 
+  // These custom arguments should be right before the object file at the end
+  context.Args.AddAllArgs(Arguments, options::OPT_Xlinker);
+  context.Args.AddAllArgs(Arguments, options::OPT_linker_option_Group);
```
[the remark this answers is not in the document]
Comment (keith) [reply lib/Driver/UnixToolChains.cpp]: Done!

File: lib/Driver/UnixToolChains.cpp [hunk 2]
```diff
@@ -325,6 +322,10 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
     Arguments.push_back("-v");
   }
 
+  // These custom arguments should be right before the object file at the end
```
[the remark this answers is not in the document]
Comment (keith) [reply lib/Driver/UnixToolChains.cpp]: Done!

File: lib/Driver/UnixToolChains.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jrose-apple) [reply lib/Driver/UnixToolChains.cpp]: Can you add a test for this as well? That would just be an invocation that uses both `-L` and `-Xlinker`.

[the remark this answers is not in the document]
Comment (keith) [reply lib/Driver/UnixToolChains.cpp]: Added, renamed the tests too since it's now not just about Xlinker

**Review by jrose-apple — APPROVED** [review]
Awesome, thanks Keith!

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Comment by keith** [comment]
@jrose-apple I think this one's ready to go 🙏

**Comment by jrose-apple** [comment]
Oops, sorry, forgot to assign it to myself so I wouldn't lose track of it.

**Comment by keith** [comment]
No worries, thanks!

**Pull request merged**